### ISSUE TITLE
Check out git commits by SHA, not by branch name

### ISSUE THREAD

**Issue by immerrr** [Shippable/support#4495]
I have just hit a weird corner case.

- I had a PR with a branch, for which a build was scheduled, but got stuck waiting for a free node
- the PR was minor, it passed code review, tests succeeded locally and it was blocking other activities, so I merged the PR without waiting for the build result
- I have a habit of deleting branches on GitHub once the corresponding PR is merged, so I proceeded to delete the branch
- At some point later in the future the build got unblocked and tried to fetch the specified branch, but failed, because the branch was already deleted (but the commit was still there).

Arguably, if I didn't have the habit of deleting branches it wouldn't fail, but there's another bug regarding this behaviour if you don't have the setting to cancel same-branch builds when a new commit arrives:

- suppose there's build for commit A for branch X that's waiting for a node to run
- suppose you add commit B to the same branch X, it schedules a new build, but doesn't cancel the build for A
- at some point a node gets freed and the build for A starts, checks out the branch, but essentially runs the build for commit B creating an inconsistency in results.

TL;DR. sources from git should be checked out via SHA-1 hash, not via branch name. I'd suggest  something like `git checkout -B <branch_name> <sha_1>` instead of `git checkout -f <branch_name>` that's used right now.

**Comment by rageshkrishna** [comment]
If I remember correctly, we used to check out by SHA many years ago. We then got complaints from several open source projects that their PR builds were failing because it was routine for them to force-push commits to their PR branches. We switched the behaviour because these folks were having the opposite problem  -- the branch still exists, but the commit does not :)

To be clear, we do this only for PR builds. Actual commits are always checked out by SHA. I'm not sure if there is any value in building earlier commits in a PR. Isn't the more important question whether the tip of the PR will build successfully when merged?

Since checking out by branch was done for a reason, and it has been the established behaviour on Shippable for many years now, it's unlikely that we will be able to change this now without breaking the workflow for many projects.
